packages/orderbook/src/order_provider/mesh_order_provider.ts

```diff
@@ -0,0 +1,207 @@
+            const assetDataB: Asset = { assetData: assetB, minAmount, maxAmount, precision };
+            assetPairsItems.push({ assetDataA, assetDataB });
+            assetPairsItems.push({ assetDataA: assetDataB, assetDataB: assetDataA });
+        }
+        return assetPairsItems;
+    }
+
+    /**
+     * Retrieves the entire orderbook for the asset pair. If websocket is available
+     * a subscription will be created to keep these orders up to date. If no websocket
+     * connection is available calling this function will fetch orders from the SRA endpoint.
+     * @param makerAssetData the Maker Asset Data
+     * @param takerAssetData the Taker Asset Data
+     */
+    public async createSubscriptionForAssetPairAsync(_makerAssetData: string, _takerAssetData: string): Promise<void> {
+        // Create the subscription first to get any updates during waiting for the request
+        await this._initializeIfRequiredAsync();
+    }
+
+    /**
+     * Submits the SignedOrder to the Mesh node
+     * @param orders the set of signed orders to add
+     */
+    public async addOrdersAsync(orders: SignedOrder[]): Promise<AcceptedRejectedOrders> {
+        const { accepted, rejected } = await utils.attemptAsync(() => this._wsClient.addOrdersAsync(orders));
+        return {
+            accepted: accepted.map(o => o.signedOrder),
+            rejected: rejected.map(o => ({ order: o.signedOrder, message: o.status.message })),
+        };
+    }
+
+    /**
+     * Destroys the order provider, removing any subscriptions
+     */
+    public async destroyAsync(): Promise<void> {
+        this._wsClient.destroy();
+    }
+
+    /**
+     * Creates the order subscription unless one already exists. If one does not exist
+     * it also handles the reconnection logic.
+     */
+    private async _initializeIfRequiredAsync(): Promise<void> {
+        if (this._wsSubscriptionId) {
+            return;
+        }
+        this._wsSubscriptionId = await this._wsClient.subscribeToOrdersAsync(this._handleOrderUpdates.bind(this));
+        await this._fetchOrdersAndStoreAsync();
+        // On Reconnnect sync all of the orders currently stored
+        this._wsClient.onReconnected(() => {
+            void this._syncOrdersInOrderStoreAsync();
+        });
+    }
+
+    /**
+     * Syncs the orders currently stored in the OrderStore. This is used when the connection to mesh
+     * has reconnected. During this outage there are missed OrderEvents so all orders are re-validated
+     * for every known asset pair.
+     */
+    private async _syncOrdersInOrderStoreAsync(): Promise<void> {
+        for (const assetPairKey of this._orderStore.keys()) {
+            const currentOrders = this._orderStore.getOrderSetForAssetPair(assetPairKey);
+            const { rejected } = await utils.attemptAsync(() =>
+                this._wsClient.addOrdersAsync(Array.from(currentOrders.values()).map(o => o.order)),
+            );
+            // Remove any rejected orders
```

Comment:
You cannot simply remove all rejected orders. For `EthRPCRequestFailed`, `CoordinatorRequestFailed`, `CoordinatorEndpointNotFound`, `InternalError`, ideally we'd retry adding them. We can leave it at is, but can you flesh out this comment to explain that this is a `// HACK`?

Reply:
Retry logic should live inside of `wsClient` rather than be exposed to the caller to determine on what should be retryable. 

`CoordinatorEndpointNotFound` is a network error or is this actually retryable?